File: solutions/go/perfect-numbers/1/perfect_numbers.go
package perfect

import (
	"errors"
	"math"
)

type Classification uint8

const (
	ClassificationPerfect Classification = iota
	ClassificationAbundant
	ClassificationDeficient
)

var ErrOnlyPositive = errors.New("only positive numbers possible")

func Classify(number int64) (Classification, error) {
	if number < 1 {
		return ClassificationDeficient, ErrOnlyPositive
	}

	var aliquotSum int64
	for divisor := int64(1); divisor <= int64(math.Floor(float64(number)/2)); divisor++ {
		if number%divisor == 0 {
			aliquotSum += divisor
		}
	}
	if aliquotSum == number {
		return ClassificationPerfect, nil
	}
	if aliquotSum > number {
		return ClassificationAbundant, nil
	}
	return ClassificationDeficient, nil
}

File: solutions/javascript/perfect-numbers/1/perfect-numbers.js
/**
 * Classify a given number as perfect, abundant, or deficient.
 * @param {number} number - The number to be classified. Must be a natural number.
 * @throws Will throw an error if the number is less than 1.
 * @returns {string} The classification of the number ('perfect', 'abundant', or 'deficient').
 */
export function classify(number) {
  if (number < 1) {
    throw new Error('Classification is only possible for natural numbers.');
  }

  const sumOfFactors = Array.from({ length: Math.floor(number / 2) }, (_, i) => i + 1)
    .filter(i => number % i === 0)
    .reduce((acc, factor) => acc + factor, 0);

  if (sumOfFactors === number) return 'perfect';
  if (sumOfFactors < number) return 'deficient';
  return 'abundant';
};
